### ISSUE TITLE
Need specification for "resource set", R

### ISSUE THREAD

**Issue by grondo** [flux-framework/rfc#109]
This issue is being opened to start a discussion on the use cases, API, and/or specification for the *_R_* as in RFC 15. *R* is the serialized version of any resource set, and is presumably produced by the serializer described in [RFC4](https://github.com/flux-framework/rfc/blob/master/spec_4.adoc#graph), consumed by the resource service in an instance as configuration, and used by the IMP and job shell to determine shape of containment and local resource slots.

In essence, the *R* format will be the way composite resource and resource configuration information will be transmitted to and from instances of Flux.

Ideally, the purpose of this issue is to determine the format of *R* such that a new RFC could be drafted.

To get the discussion started, here are some high level requirements and use cases for *R*:

 - *R* should act as resource configuration input to an instance,
   therefore it may be that configuration of even the system instance
   is written in *R* spec, or the configuration language (RDL?) generates *R*.
   (in fact, one use case might be to directly generate R from hwloc data)

 - Execution service in an instance needs to be able to generate
   *R<sub>local<sub>* from *R* for each rank. So given a rank or even generic
   "resource vertex", there should be a function to generate an
    *R<sub>n<sub>* from *R*, where *R<sub>n<sub>* is a hierarchical subset of *R*.

 - The containment plugins in the IMP will need to query *R<sub>local<sub>*
   for the list of local resources of given type or types on which
   the containment plugins operate. For instance, a memory plugin
   will need to determine the amount and location of RAM contained
   in *R<sub>local<sub>* in order to set up memcg limits. Similarly a Socket/CPU
   plugin would need to iterate over or query the list of *local*
   sockets/cores in *R<sub>local<local>* to add these to the cgroup.

 - The job shell will use jobspec+*R* to determine the local
   'task slots' that map to commands in the 'tasks' section.

Dependency management here might get challenging. The IMP is a user of *R<sub>n<sub>*, but we want to ideally eliminate dependencies in the flux-security project on other flux-framework projects. Possible approaches here might include:
1. The IMP could take a subset of the *R* specification, simple enough to parse with its own parser and offer some simplified interface to IMP plugins that do containment. Containment plugins probably only really need to get a list of local resources from *R*, as long as the logical IDs match the actual system logical IDs. The IMP's interface to R could later be expanded to offer higher level functionality for advanced containment (though I don't have any use cases that I can think of here)
2. Alternately, the IMP itself could treat *R* as opaque data, passed to plugins. The plugins would then have a dependency on some library from a *system-installed* flux core or sched project.

**Comment by lipari** [comment]
Nice start, @grondo.  Just wanted to add some distinctions... The R will contain node-resident resources that will become *R<sub>local<local>* as well as ancillary resources like switches, racks, licenses, bandwidth and power.  For resources like a rack or a switch, the scheduler may schedule jobs that are distributed to specific racks or switches.  However, one could imagine that the allocated rack or switch in this case remains in the scheduler's domain and does not get included in the R that is passed to the system instance (as embodied by a collection of brokers).

On the other hand, for resources for which we will have controls (e.g., license managers), these resources would become part of R but would only be relevant to specific agents like license managers, and not the brokers (unless we have a broker devoted to controlling a license manager).

Power and network bandwidth could fall into either of these two cases depending on whether a throttle was available to limit power or bandwidth serving the allocated resources.

**Comment by grondo** [comment]
Thanks, @lipari! You bring up some good points.

> On the other hand, for resources for which we will have controls (e.g., license managers), these resources would become part of R but would only be relevant to specific agents like license managers, and not the brokers (unless we have a broker devoted to controlling a license manager).

These generic "global" job resources like licenses, burst buffer storage space, bandwidth, etc will have to be passed in to *some* sort of containment management, in case there is some action required to give access to the licenses, or reserve space, etc. My thought is that these would be included in *R<sub>local<sub>*, and then the container plugin specific to that resource type would be able to decide how to contain or make available that specific resource. For example, a simple approach would be to have the plugins on the first node of a job operate on these resources.

What remains to be decided is how these resources do get included in *R<sub>n<sub>* for each IMP *n*. We may have to put some tag on these kind of global resources so they are automatically included in any *R<sub>n<sub>*

I would also argue that *R<sub>local<sub>* for any IMP should include not only the resource vertex(es) on which the IMP will be run, but also all the parents of the vertex up to the root (in the hierarchical resource tree). This will give the IMP containment plugins a bit more information about where they are running in the global hierarchy, which could be useful, and also allows us to keep global resources discussed above in their proper place in any hierarchy.

E.g. instead of a simple *R<sub>local<sub>* like `socket[0]->core[0-3]` for an IMP managing a single socket, you might instead have `llnl->cluster[5](name=hype)->node[113](name=hype113)->socket[0]->core[0-3]`

This probably won't be used at first

**Comment by morrone** [comment]
> My thought is that these would be included in Rlocal, and then the container plugin specific to that resource type would be able to decide how to contain or make available that specific resource.

> I would also argue that Rlocal for any IMP should include not only the resource vertex(es) on which the IMP will be run, but also all the parents of the vertex up to the root (in the hierarchical resource tree). 

I'm worried that these new parts of the plan are undermining the original need/goal for R<sub>local<sub>.  If the R<sub>local<sub> contains resource types that the IMP doesn't directly control, then we are back to the original situation we had with a single large R.  The IMP now needs to parse out and identify only its locally relevant resource types from a larger tree of information.  Rather than going that route, I wonder if it wouldn't make more sense to just going back to having a single complete R document that is sent everywhere including the IMP (although this time around we are deciding that the complete R no longer needs to be signed).

In other words, if the IMP needs to parse out potentially extraneous information from R<sub>local<sub> and find the point where the information begins to align with its local resources, then it could do that just as easily from a large R.  What is the value then of R<sub>local<sub>?

**Comment by grondo** [comment]
> In other words, if the IMP needs to parse out potentially extraneous information from Rlocal and find the point where the information begins to align with its local resources, then it could do that just as easily from a large R. What is the value then of Rlocal?

*R<sub>local<sub>* allows the instance that is starting the IMP to control the shape of the container under which the IMP will execute the job shell, instead of relying on the IMP to make that decision, when it doesn't have or need the necessary data to make the correct decision about what goes in the local "container". 

I guess where you might differ in opinion is whether the parents of a resource are part of that conceptual container. I tend to think a container that is just "cpu0" doesn't make any sense, you need `node->socket[0]->cpu[0]` *at least* to resolve the container. Taking that idea a bit further, in our resource model, `node0` is not a valid container either, you need `llnl->hype->node[0]`. The IMP won't have any containment plugins that try to operate on resource type "datacenter", "node", or "switch", etc., so the extra resources will be safely ignored. However, if a containment plugin happens to need this information, it may be at least able to get it. (location of off-node resources is the main use case I'm considering now)

Another benefit of *R<sub>local<sub>* is potentially eliminating dependence on flux-sched or flux-core provided resource query language that might be required to perform the intersection between local resources and global R (though even if you had this support for the IMP, I'm not convinced the IMP alone could make the right decision here). To realize this particular goal, *R<sub>local<sub>* will need to be simple enough that the IMP or its plugins could parse it easily themselves.

**Comment by grondo** [comment]
A specific case where *R<sub>local<sub>* might be required is if an instance, for testing or other good reason, would like to start more than one IMP per broker. To do this, the instance would break up local resources into multiple *R<sub>local<sub>* and pass to each IMP. I don't see how it would be possible if the instance passed the global *R* to each IMP.

BTW, I was mainly taking a long-term view on inclusion of parent resources in *R<sub>local<sub>*, and as long as it is possible to add that support in at a future date, I'm ok with leaving it out for now. I think off-node resources like burst-buffer space and licenses could be handled by including enough metadata in those resources included in *R<sub>local<sub>* such that a plugin could know exactly *which* licenses it was operating on, or which burst buffers it was reserving space in... etc.

**Comment by morrone** [comment]
> A specific case where Rlocal might be required is if an instance, for testing or other good reason, would like to start more than one IMP per broker.

Ah.  I thought we agreed that we weren't doing that, that instead the IMP always controls all resources on the node and leaves further resource masking to the jobshell.  But if we are reversing that I suppose that is fine.

> I think off-node resources like burst-buffer space and licenses could be handled by including enough metadata in those resources included in Rlocal such that a plugin could know exactly which licenses it was operating on, or which burst buffers it was reserving space in... etc.

If we think that the IMP is going to need to know about global resources, then again I think I'm back to thinking we should just send the full R.  We can always add annotations to that if we ever want to run multiple IMPs per node.  Adding global resources to R<sub>local</sub> makes the name a bit of a misnomer. :)

**Comment by morrone** [comment]
> Another benefit of Rlocal is potentially eliminating dependence on flux-sched or flux-core provided resource query language that might be required to perform the intersection between local resources and global R

The IMP reading a global R does not necessarily imply that the IMP must use a "resource query language".  The IMP just needs a parser.  The parser for R and Rlocal, if Rlocal is allowed to contain global resources, will be very nearly identical I think.  We can choose to implement the parser twice or cut-and-paste it into the IMP if we want to keep it separate.

**Comment by grondo** [comment]
> If we think that the IMP is going to need to know about global resources, then again I think I'm back to thinking we should just send the full R. We can always add annotations to that if we ever want to run multiple IMPs per node. Adding global resources to Rlocal makes the name a bit of a misnomer. :)

Yeah, I completely understand your sentiment. I'm fine with leaving *R<sub>local<sub>* with only "local" resources (whatever "local" may mean), but then we have no proposed method to handle off-node resources (since IMP can only run at most within a node).

**Comment by morrone** [comment]
I thought we had talked about that.  I think that the "execution management" module, or whatever we are calling it now, would handle off-compute-node resource setup before launching remote execution and the IMPs.  There would be plugins into that module that can instantiate the various resources that people come up with.

**Comment by grondo** [comment]
> The IMP reading a global R does not necessarily imply that the IMP must use a "resource query language". The IMP just needs a parser. The parser for R and Rlocal, if Rlocal is allowed to contain global resources, will be very nearly identical I think. We can choose to implement the parser twice or cut-and-paste it into the IMP if we want to keep it separate.

The IMP will need to parse R, but then how does it complete the intersection between local available resources and R. It would need to generate an R' from hwloc or some other local HW query code, then take the intersection of R' and R.

If only *R<sub>local<sub>* is sent to the IMP it doesn't need to read local HW configuration, it doesn't need to generate a second R' from from that information, and it doesn't need to do the work of the intersection. So that feels like quite a bit of code saved from a security significant piece of software.

**Comment by grondo** [comment]
> I thought we had talked about that. I think that the "execution management" module, or whatever we are calling it now, would handle off-compute-node resource setup before launching remote execution and the IMPs. There would be plugins into that module that can instantiate the various resources that people come up with.

That could work but the instance doesn't have any privilege except through the IMP. Is it a requirement that all off node resources don't require privilege to access (this is possible, I just didn't think of that way before)?

**Comment by morrone** [comment]
> The IMP will need to parse R, but then how does it complete the intersection between local available resources and R. It would need to generate an R' from hwloc or some other local HW query code, then take the intersection of R' and R.

I think it is actually a lot simpler than that if there is just a single IMP per node.  It just walks the tree of data in R, and looks up each resource it sees in its internal table "oh!  that belongs to me, I'll note that", "nope that doesn't belong to me, skip it".  There is no complicated intersection needed, really.

**Comment by grondo** [comment]
> There is no complicated intersection needed, really.

Ok, I guess I couldn't visualize how to make it quite that simple.

Whereas with *R<sub>local<sub>* The IMP would walk each *type* of resource for which it has a containment plugin and hand the list of those resources in *R<sub>local<sub>* to the plugin (or alternately each plugin could generate the list itself(. No comparisons needed at all. Since there won't be containment plugins for "node" "switch" "datacenter", and other resources, those would be safely ignored if they were there at all.

**Comment by morrone** [comment]
> That could work but the instance doesn't have any privilege except through the IMP. Is it a requirement that all off node resources don't require privilege to access (this is possible, I just didn't think of that way before)?

It would be preferable when possible.  But it can be handled on a case-by-case basis.

Doing this through the IMP could potentially introduce a fair bit of complexity.  We might be back to needing a away to track authority back through multiple levels of flux instances.  We were able to avoid that when the IMP was constrained to dealing with resources inside of its local node's container.

**Comment by morrone** [comment]
> Whereas with Rlocal The IMP would walk each type of resource for which it has a containment plugin and hand the list of those resources in Rlocal to the plugin

Yeah, actually it is slightly more complicated than I stated, but not much.  Actually, in each case where it finds a resource it owns, it needs remember that AND all of the resources under it in the tree.  But that is still pretty straight forward I think.

I don't think the IMP can only look at types even for an Rlocal.  It needs to look at either names or counts too.  Flux allows two approaches to something like a "socket".  We can either represent all of the sockets on a node as a single resource vertex and use the count within that vertex to represent all of the sockets, or we can have a resource vertex with a name/id/uuid/whatever for_each_ of the sockets.

In the former case, with individual resource vertices, the scheduler picks the exact resources, and the IMP just needs to carry out the instructions.  In the latter, with numbered resources, the IMP needs to be more aware of what is happing with allocations on the node (for instance, if nodes are shared).  But actually, I'm not sure that the IMP can read the scheduler's mind enough to always make the same selection pattern...and that could lead to resources being shared (sockets/cores) that the scheduler could have avoided.  So actually, for things like sockets and cores I suspect that we will always use separate resource vertices.

> Since there won't be containment plugins for "node" "switch" "datacenter", and other resources, those would be safely ignored if they were there at all.

I think is reasonable for the IMP to know what node it is on.  And once it knows that, it will be fairly easy to pick out its own resources from the global R.

**Comment by grondo** [comment]
> Doing this through the IMP could potentially introduce a fair bit of complexity. We might be back to needing a away to track authority back through multiple levels of flux instances. We were able to avoid that when the IMP was constrained to dealing with resources inside of its local node's container.

I'm not sure that is required, but either way you need to verify ownership of the resource whether it is through an IMP plugin or a plugin in the execution system.

One example I can think of is if you restricted access to a license server through some sort of iptables rules. For jobs granted access to the license server you would have to allow iptables rules to be modified in the network namespace of the job. I can envision how this could be done with an IMP plugin, but not at all what kind of system you'd need if you tried to do it through a plugin in the unprivileged execution modules. Besides, the containers can't exist until the IMP runs so there is an ordering problem  there. 

For this case would you consider licenses a "local" resource, or perhaps rename them network access tokens or something? (If that is the case then I could see that this scheme could work)

**Comment by grondo** [comment]
> I think is reasonable for the IMP to know what node it is on. And once it knows that, it will be fairly easy to pick out its own resources from the global R.

Ok, but why? I guess it is immaterial if we pass R or *R<sub>local<sub>* since the IMP always filters R to *R<sub>local<sub>* anyway (it will just be a noop in the second case). The question I keep struggling with is why you'd want to do that work in your privileged process if you don't have to?

> I don't think the IMP can only look at types even for an Rlocal. It needs to look at either names or counts too. 

Yes, that is what I meant. Each containment plugin will need to know the list and count of types (especially for RAM) of each of the resource types it knows how to deal with.

**Comment by grondo** [comment]
Here's my proposal for a simplified *R<sub>local<sub>* to satisfy near-term milestones, if it is acceptable that *R* and *R<sub>local<sub>* have different specifications.

The simplified *R<sub>local<sub>* as input to the IMP will be a JSON document with a list of resource types for which the IMP should create a container. The IMP will support a list of plugins which operate on one or more of each type, and access the *R<sub>local<sub>* directly to determine the parameters of the containers they can create. e.g. a memory and cpu "cgroups" container would read the "sockets" "cpus" and "memory" fields of the *R<sub>local<sub>* dictionary and add cpus and mems to a cpuset cgroup, and constrain memory with a memory cgroup.

The format of *R<sub>local<sub>* might look (very roughly), something like:
```json
{
 "cpu": { "list": [ 0, 1, 2, 3] }, "count": 4 },
 "socket": { "list": [ 0 ],  "count": 1 },
 "memory": { "count": 1024, "units": "MB" },
}
```

This is just off the cuff so there may be missing fields, but is meant to give a general idea.

If the IMP must take the full R as input, then I'd suggest a plugin to the IMP, provided by the instance, would generate this format as input to the IMP containment plugin infrastructure. That would further require that the IMP operate in privilege separation mode so that the plugin operating on R runs with permissions of the instance owner. This would avoid copy-and-paste parsing code between flux-framework projects, and allow a single, system-installed version of flux-security to work with multiple versions of other Flux projects which may generate R with different formats or capabilities.

**Comment by morrone** [comment]
> The question I keep struggling with is why you'd want to do that work in your privileged process if you don't have to?

Me too!  Which is why adding non-local things to Rlocal that need to be skipped seems like exactly the same kind of processing in a privileged process that we were trying to avoid by having Rlocal in the first place.  I don't see might difference between skipping 3 things or skipping 1000.  The parsing code needs to be rock solid in either case, so the implementation work seems the same.  But if they are the same, then why put the extra effort into implementing Rlocal in the first place.

> One example I can think of is if you restricted access to a license server through some sort of iptables rules.

That is interesting...is that how it usually works?  I was thinking more that some number of floating licenses would be allocated to an entire job, and they could choose to use them where they like.  But I guess it depends on the sophistication of the particular license server.

**Comment by grondo** [comment]
> That is interesting...is that how it usually works? I was thinking more that some number of floating licenses would be allocated to an entire job, and they could choose to use them where they like. But I guess it depends on the sophistication of the particular license server.

I actually don't know, but it was a real proposal from somewhere (but not sure if it was ever implemented).

> Me too! Which is why adding non-local things to Rlocal that need to be skipped seems like exactly the same kind of processing in a privileged process that we were trying to avoid by having Rlocal in the first place. I don't see might difference between skipping 3 things or skipping 1000. The parsing code needs to be rock solid in either case, so the implementation work seems the same. But if they are the same, then why put the extra effort into implementing Rlocal in the first place.

You do make good points.
I guess you are implementing Rlocal in either case, it is just a case of where it is generated. I say you have my consensus that non-local things should not go into Rlocal, and lets leave it at that for now.
However, also take a look at my proposal above at possibly using instance provided code to parse R from within the IMP by running it in an unprivileged child.

Either way, I don't think we're closing the door on including either more or less resources in the R in future implementations, so I'm thinking we can move on for now...

**Comment by morrone** [comment]
> However, also take a look at my proposal above at possibly using instance provided code to parse R from within the IMP by running it in an unprivileged child.

The devil is in the details there.  At some point the IMP needs to ingest resource data in some form from an untrusted source.  It could either be by parsing a resource document itself, or from parsing the output of something that parses the resource document.  But at some point it always needs to be defensive and validate its input.  I would need to know more to evaluate it, I think.

I'm not really clear why parsing R is more seems more scary than parsing Rlocal, or why it would necessarily need to be handled separately through plugins and/or resource separation.  I think R would be handled by an internal parser in the IMP exactly the same way that Rlocal is being proposed to be handled.

**Comment by grondo** [comment]
> I'm not really clear why parsing R is more seems more scary than parsing Rlocal, or why it would necessarily need to be handled separately through plugins and/or resource separation. I think R would be handled by an internal parser in the IMP exactly the same way that Rlocal is being proposed to be handled.

Ok, again I understand your point.

I guess I'm arguing that *R<sub>local<sub>* as used internally by the IMP is a different, much simpler, format than *R* (sorry, maybe it shouldn't be called Rlocal anymore?). The amount of code being used would therefore be less, and therefore, provably less bugs.

The Rlocal format could evolve much more slowly than R, though I admit it hasn't been proven that R will change at any kind of pace that would require frequent updates to flux-security project, so perhaps a weaker argument here.

Also it just kind of seems to make sense to send less data to the IMP, even though this doesn't have a security argument. For a job with 1000 cores on 1000 nodes, R is potentially 1000x the size of Rlocal....

**Comment by lipari** [comment]
>  Also it just kind of seems to make sense to send less data to the IMP, even though this doesn't have a security argument. For a job with 1000 cores on 1000 nodes, R is potentially 1000x the size of Rlocal....

Totally agree.

**Comment by grondo** [comment]
To summarize, I think consensus here is that *R<sub>local<sub>* should contain *only* node local resources, but that it is still useful to send only a subset of *R*, perhaps in a simpler format, as input to the IMP.

As far as the topic of this issue, which is the specification of *R*, I don't think that changes much. We still need to be able to generate some *R<sub>local<sub>* from *R*, and therefore we'll need some kind of format of *R* that allows this within in an instance.

**Comment by dongahn** [comment]
> To summarize, I think consensus here is that Rlocal should contain only node local resources, but that it is still useful to send only a subset of R, perhaps in a simpler format, as input to the IMP.

IMHO this approach is sound.

As I already discussed this w/ @grondo, extracting "*R<sub>local<sub>*" should be fully distributed so that a centralized component doesn't become a scalability bottleneck.

For needing to fetch *R* by many execution service modules to extract "*R<sub>local<sub>*" , I believe this should be scalable as this would essentially have performance complexity of a broadcast... At some point, we may want to measure this though..

Sorry if this was too obvious.

**Comment by dongahn** [comment]
I would like to have a bit more discussions on the main topic of this issue: **R** format. As @grondo nicely captured this in the beginning, **R** will serve as the input and output of a range of components. For example, it will be the input to the `resource` service (resource selection part of scheduling) in a nested instance, as well as `remote execution` service, and job shell. Similarly, it will be the output of the `resource` service, and also of other related services, utilities, and even manual effort, e.g., : 
- hwloc (probably via other services that help with formatting);
- [GRUG](https://github.com/dongahn/resource-query#generating-resources-using-graphml-grug) through `resource` service;
- RDL through `resrc`;

It seems an important decision we can all benefit from at this point would be whether we want to spec out a common **R** format or going with an opaque  approach with just common abstractions on it agreed upon.

W/o looking at this too closely, if we go with a graph format  with an optional ability to annotate extra  information on resource vertex and edge (i.e., the concepts described in RFC4), we should be able to describe the format captured in all of the above use cases, and I can contribute to that effort based on my current`resource-query` experience.

But a part of me is also saying whether this kind of rigor on the format is necessary at this point. 

Another approach can be for each of format variants, we require a library to expose a set of common operations including "reader" and "writer."

The former would a bit more rigorous but at the same time it can be a bit more time consuming. But maybe something that we ought to do anyway.

There is also the third possibility which is to start to go with the opaque approach above but as we reach agreements on the common abstractions, we will know the requirement on the format better. And at that point, we can formalize the format. It seems we will have to write libraries that expose those abstractions  around the **R** anyway, this also is not a bad idea IMHO either.

Thoughts?

**Comment by grondo** [comment]
Thanks @dongahn. Some very good thoughts above.

Another element we need to keep in mind is how various components of flux would manage dependencies of interpreting and managing *R*. Ideally perhaps, *R* format would be supported directly by flux-core, so that the execution system, which depends on it, can be tested stand-alone. However, this approach might lead to a lowest-common denominator format, which may not support the needs of advanced resource services and/or schedulers.

An argument might also be made the the *R* format is solely the domain of a resource service, and the *R* interpreter and generator should therefore by offered by that service, though that would leave the execution service dependent on resource services being installed, which might not be what we want.

Another approach would be to have the *R* spec opaque as you said, with each type supplying a corresponding API that satisfies the requirements of all use cases outside of resource service internals. Somehow the required implementation would be encoded in *R* itself and the correct implementation loaded at runtime.

One more idea would be to have a very basic *R* specification, but allow a section for "extensions" which might be ignored by most components, but used for any extra information needed by the resource service itself. The base *R* spec might not even need an API if it was simple enough, thereby removing the pain of deciding where the dependent libraries might live.

**Comment by dongahn** [comment]
Of four possibilities, the last two seems most attractive from my perspective. A bit on the 4th option:

> One more idea would be to have a very basic R specification, but allow a section for "extensions" which might be ignored by most components, but used for any extra information needed by the resource service itself. The base R spec might not even need an API if it was simple enough, thereby removing the pain of deciding where the dependent libraries might live.

This is a very interesting idea from my perspective @grondo. I consider the resource representation needed by `flux-core` elements as a "proper" subset of the representation needed by `resource` so this can work out nicely if we can reasonable separate out the baseline from the extension. I don't know if "section" is the right construct but I got the idea.

Why don't I put a few examples of the graph representations I plan to use by `resource` and see what belongs to the baseline and what belongs to the core elements, and see if these are easily separable. I will use GraphML but other markup language capable of graph can do.

**Comment by grondo** [comment]
sounds good @dongahn! Thanks!

**Comment by dongahn** [comment]
Here is an GraphML example that describes an **R** such that it has 1 cluster with 1 rack with 1 node with 2 sockets each with 2 cores/1 GPU/memory. This was actually emitted from my `resource-query` utility. I will explain parts of this a bit in a separate posting as relevant to this ticket.

But the general description on GraphML itself can be found in something like [this](https://cs.brown.edu/%7Ert/gdhandbook/chapters/graphml.pdf). The reason that I used GraphML is that, our resource model is essentially is a graph and as such I didn't think I needed to reinvent the wheel with other markup languages. Plus, there are already plenty of GraphML readers and writers out there including Boost Graph Library. 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<graphml xmlns="http://graphml.graphdrawing.org/xmlns" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://graphml.graphdrawing.org/xmlns http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd">
  <key id="basename" for="node" attr.name="basename" attr.type="string" />
  <key id="id" for="node" attr.name="id" attr.type="long" />
  <key id="subsystems" for="node" attr.name="member_of" attr.type="string" />
  <key id="esubsystems" for="edge" attr.name="member_of" attr.type="string" />
  <key id="name" for="node" attr.name="name" attr.type="string" />
  <key id="paths" for="node" attr.name="paths" attr.type="string" />
  <key id="props" for="node" attr.name="props" attr.type="string" />
  <key id="size" for="node" attr.name="size" attr.type="int" />
  <key id="type" for="node" attr.name="type" attr.type="string" />
  <key id="unit" for="node" attr.name="unit" attr.type="string" />
  <graph id="G" edgedefault="directed" parse.nodeids="canonical" parse.edgeids="canonical" parse.order="nodesfirst">
    <node id="n0">
      <data key="basename">small</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">small0</data>
      <data key="paths">{containment: &quot;/small0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">cluster</data>
      <data key="unit"></data>
    </node>
    <node id="n1">
      <data key="basename">rack</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">rack0</data>
      <data key="paths">{containment: &quot;/small0/rack0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">rack</data>
      <data key="unit"></data>
    </node>
    <node id="n2">
      <data key="basename">node</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">node0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">node</data>
      <data key="unit"></data>
    </node>
    <node id="n3">
      <data key="basename">socket</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">socket0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">socket</data>
      <data key="unit"></data>
    </node>
    <node id="n4">
      <data key="basename">socket</data>
      <data key="id">1</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">socket1</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket1&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">socket</data>
      <data key="unit"></data>
    </node>
    <node id="n5">
      <data key="basename">core</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">core0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0/core0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">core</data>
      <data key="unit"></data>
    </node>
    <node id="n6">
      <data key="basename">core</data>
      <data key="id">1</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">core1</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0/core1&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">core</data>
      <data key="unit"></data>
    </node>
    <node id="n7">
      <data key="basename">core</data>
      <data key="id">2</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">core2</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket1/core2&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">core</data>
      <data key="unit"></data>
    </node>
    <node id="n8">
      <data key="basename">core</data>
      <data key="id">3</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">core3</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket1/core3&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">core</data>
      <data key="unit"></data>
    </node>
    <node id="n9">
      <data key="basename">gpu</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">gpu0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0/gpu0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">gpu</data>
      <data key="unit"></data>
    </node>
    <node id="n10">
      <data key="basename">gpu</data>
      <data key="id">1</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">gpu1</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket1/gpu1&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">gpu</data>
      <data key="unit"></data>
    </node>
    <node id="n11">
      <data key="basename">memory</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">memory0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0/memory0&quot;}</data>
      <data key="props"></data>
      <data key="size">4</data>
      <data key="type">memory</data>
      <data key="unit"></data>
    </node>
    <node id="n12">
      <data key="basename">memory</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">memory0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket1/memory0&quot;}</data>
      <data key="props"></data>
      <data key="size">4</data>
      <data key="type">memory</data>
      <data key="unit"></data>
    </node>
    <edge id="e0" source="n0" target="n1">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e1" source="n1" target="n0">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e2" source="n1" target="n2">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e3" source="n2" target="n1">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e4" source="n2" target="n3">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e5" source="n2" target="n4">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e6" source="n3" target="n2">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e7" source="n3" target="n5">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e8" source="n3" target="n6">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e9" source="n3" target="n9">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e10" source="n3" target="n11">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e11" source="n4" target="n2">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e12" source="n4" target="n7">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e13" source="n4" target="n8">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e14" source="n4" target="n10">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e15" source="n4" target="n12">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
    <edge id="e16" source="n5" target="n3">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e17" source="n6" target="n3">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e18" source="n7" target="n4">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e19" source="n8" target="n4">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e20" source="n9" target="n3">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e21" source="n10" target="n4">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e22" source="n11" target="n3">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
    <edge id="e23" source="n12" target="n4">
      <data key="esubsystems">{containment: &quot;in&quot;}</data>
    </edge>
  </graph>
</graphml>
```

**Comment by dongahn** [comment]
As you can see from:

```xml
  <key id="basename" for="node" attr.name="basename" attr.type="string" />
  <key id="id" for="node" attr.name="id" attr.type="long" />
  <key id="subsystems" for="node" attr.name="member_of" attr.type="string" />
  <key id="esubsystems" for="edge" attr.name="member_of" attr.type="string" />
  <key id="name" for="node" attr.name="name" attr.type="string" />
  <key id="paths" for="node" attr.name="paths" attr.type="string" />
  <key id="props" for="node" attr.name="props" attr.type="string" />
  <key id="size" for="node" attr.name="size" attr.type="int" />
  <key id="type" for="node" attr.name="type" attr.type="string" />
  <key id="unit" for="node" attr.name="unit" attr.type="string" />
```

a resource pool vertex contains 9 ~ 10 base fields (`uuid` omitted for now):
- `basename`
- `id`
- `name`
- `path` (JSON dictionary string)
- `properties` (JSON dictionary string)
- `size`
- `type`
- `unit`
- `subsystems` (JSON dictionary whose keys are subsystems that this vertex belongs to)

For example, the following resource vertex construct describes  a socket resource  (`socket0`) contained within `node0`. Note that `&quot` is an xml encoding scheme for double quote (`"`):
```xml
    <node id="n3">
      <data key="basename">socket</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">socket0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">socket</data>
      <data key="unit"></data>
    </node>
```

The following vertex construct describes  a core resource  (`core0`) contained within `socket0`:
```xml
    <node id="n5">
      <data key="basename">core</data>
      <data key="id">0</data>
      <data key="subsystems">{containment: &quot;*&quot;}</data>
      <data key="name">core0</data>
      <data key="paths">{containment: &quot;/small0/rack0/node0/socket0/core0&quot;}</data>
      <data key="props"></data>
      <data key="size">1</data>
      <data key="type">core</data>
      <data key="unit"></data>
    </node>
```

Each edge construct describes a directional relationship between two resource vertices and has one data field. For example, the following describes a relational edge from `socket0` vertex (whose vertex id is `n3`) to `core0` vertex (`n5`).

```xml
    <edge id="e7" source="n3" target="n5">
      <data key="esubsystems">{containment: &quot;contains&quot;}</data>
    </edge>
```

Now, my guess is that anything but `subsystem` and `esubsystem` fields would be the baseline **R** as needed by execution systems and other generators. But we should discuss. 

In general, I believe **R** that is capable of describing a graph would be the most expressive way to support a wide range of producers and consumers. For example, a nested`resource` service can simply pass **R** from the guest KVS to `read_graphml` to populate its graph data store before starting the nested scheduling service. 

Other generators like a hwloc adapter can also easily emit its information to this format. hwloc won’t have the concept of subsystems so the `subsystem` and `esubsystem` fields would be empty, however.

One thing that bothers me a little is this encodes graph data in a bit redundant manner. But this is because the preexisting graphml writer emits the graph in this manner. Thus, this can be compressed to a good degree if we decide to enhance our own writers.

**Comment by lipari** [comment]
While GraphML makes sense as a resource representation to the scheduler, I think the resources passed to the execution engine on each node (still called *R<sub>local<local>*?) should be conveyed via hwloc's topology construct (https://www.open-mpi.org/projects/hwloc/doc/v1.2/structhwloc__obj.php).  This will facilitate the binding the execution engine needs to do, leveraging the power of the hwloc library:  https://www.open-mpi.org/projects/hwloc/doc/v1.2/group__hwlocality__cpubinding.php.  I realize that binding is not a subject of this issue - and that we may opt to use cgroups directly instead of through the hwloc library.  However, I wanted to raise the comment that the graph format for R should be different from the node-focused representation of *R<sub>local<local>*.

**Comment by dongahn** [comment]
Thanks @lipari. I believe that the stated plan is to do containment via a plugin to IMP and the binding at the jobshell level. Both will need adaptors to extract the info from the **R** to the format they require. 

@grondo's proposal to R local is up there. How jobsell will access the **R** was discussed yesterday and there is a PR #114 as a result of that.

**Comment by dongahn** [comment]
BTW, as I remember hwloc also supports reading from and writing to xml. My guess is translating the entire **R** to this xml format should be pretty straightforward.

**Comment by lipari** [comment]
Sorry, @dongahn.  I read the [PR 114](https://github.com/flux-framework/rfc/pull/114) discussion of mods to [RFC 16](https://github.com/flux-framework/rfc/blob/master/spec_16.adoc) as describing how the GraphML formatted R is represented in the KVS.  It is unclear to me where we have specified the work of translating a GraphML R to the *R<sub>local<local>* form and where the *R<sub>local<local>*'s format is defined.  I thought I read all the discussion, but I probably missed it.

**Comment by lipari** [comment]
To clarify... I think it is asking a lot of the IMP to break a GraphML-formatted R down to *R<sub>local<local>* which needs to look like an hwloc topology object - even with an adapter.  I'm questioning whether that is the best plan.

**Comment by dongahn** [comment]
We am open to alternatives, of course. 

If the node-local resource discovery speed is an issue, i suppose we can add some index information such that node-local information can be found more expediantly...

What is your alternative, @lipari?

**Comment by dongahn** [comment]
> ... translating a GraphML R to the Rlocal form and where the Rlocal's format is defined. I thought I read all the discussion, but I probably missed it.

https://github.com/flux-framework/rfc/issues/109#issuecomment-334306854 would be closest, I think.

**Comment by grondo** [comment]
Sorry, just catching up on this. 

@lipari, some clarification: the execution subsystem creates *R<sub>local<sub>*, but it is part of the enclosing instance and therefore can use instance services to create this representation. The exec service doesn't do binding, that is completed by the IMP, and therefore *R<sub>local<sub>* will be in a format defined by the IMP, likely something very simple, not GraphML.

**Comment by lipari** [comment]
> #109 (comment) would be closest, I think.

Ah, thank you for the reminder.

> What is your alternative, @lipari?

@dongahn, you are proposing GraphML to be the form of R.  And I agree that is a sound proposal for the reasons you stated.  But to be confident that GraphML is the best choice, it behooves us to consider how easily GraphML can be parsed to generate the node-focused *R<sub>local<local>*.  And I was attempting to minimize the complexity of the IMP, if it needs to generate *R<sub>local<local>*.  If this is turns out to be no problem for the IMP, then I'm fine with using GraphML to represent R.

**Comment by grondo** [comment]
>  This will facilitate the binding the execution engine needs to do, leveraging the power of the hwloc library: https://www.open-mpi.org/projects/hwloc/doc/v1.2/group__hwlocality__cpubinding.php.

The job shell might use hwloc for *binding*, but the IMP will need to create a container from *R<sub>local<sub>* and I don't think the hwloc topology format nor the library itself will be helpful there.

**Comment by grondo** [comment]
> And I was attempting to minimize the complexity of the IMP, if it needs to generate Rlocal.

I think we concluded that the most efficient approach is that *R<sub>local<sub>* is *input* to the IMP, not generated by the IMP. 

If it is later decided that the input to the IMP should be *R*, from which the IMP would generate its own *R<sub>local<sub>*, then I would expect the IMP would utilize a library from the enclosing instance to operate on *R*. 

In either case the IMP will always treat *R* as opaque and should not need to know its specific format.

**Comment by dongahn** [comment]
> And I was attempting to minimize the complexity of the IMP, if it needs to generate Rlocal. If this is turns out to be no problem for the IMP, then I'm fine with using GraphML to represent R.

I thought you had a good point in that node-local resource discovery from **R** should be efficient. I am sure we will revisit this as "optimization" topic as we get down to the implementation...

**Comment by dongahn** [comment]
To summarize the discussion points so far:
- There seem good benefits in actually defining our spec for **R** that supports baseline + extensions, and GraphML format is proposed.
- The execution engine will create *R<sub>local<sub>* and pass that as the input to the IMP
- Jobshell will get to **R** to do binding
- We may need support for efficiently discovering node-local resources

From my perspective, gathering a bit more feedback on the GraphML proposal will help.

**Comment by lipari** [comment]
Thank you for the summary, @dongahn!  It seems to me this discussion has bearing on whether to make GraphML be the format of R.

**Comment by grondo** [comment]
Thanks @dongahn.

> It seems to me this discussion has bearing on whether to make GraphML be the format of R.

Yeah, the question is whether GraphML can be the *baseline* format of **R**

> * Jobshell will get to **R** to do binding

The job shell, as well as any sub-instance started by the shell may require **R** in such a format that it can be parsed and used for binding or as configuration input without a special library from the parent, which may be running as a different user. This makes me think that a simpler "baseline format" may be required. Instance implementations may still use GraphML internally, but they would need a converter to and from the baseline format, with extra information stored in the extensions. At this point I'm not sure otherwise how the job shell and other instances running potentially different versions of Flux could reliably and safely make use of **R** directly.

**Comment by dongahn** [comment]
Just for an argument sake,

 > The job shell, as well as any sub-instance started by the shell may require R in such a format that it can be parsed 

GraphML is an XML and this can be parsed easily.

> and used for binding 

The parsed objects can be easily traversed and used for binding.

> as configuration input without a special library from the parent, which may be running as a different user. 

Again, this is just an XML and its dependency is xml not a special library from the parent instance.

> This makes me think that a simpler "baseline format" may be required.

What do you have in mind? 

> This makes me think that a simpler "baseline format" may be required. Instance implementations may still use GraphML internally, but they would need a converter to and from the baseline format, with extra information stored in the extensions.

Is the worry having a baseline section and exertion section such a way that one can only parse the baseline section for baseline operation?

> At this point I'm not sure otherwise how the job shell and other instances running potentially different versions of Flux could reliably and safely make use of R directly.

Again, GraphML is an XML; so you don't have to require a special library from the parent as far as the format is formally specified.

**Comment by grondo** [comment]
> GraphML is an XML and this can be parsed easily.

Ok, I was worried that boost::graph library or special API from parent instance would be required.

> What do you have in mind?

I don't have a good idea. But simple baseline format could be a yaml representation similar to that used in jobspec. This would cut the library requirements down from xml+yaml to just yaml. This is just one idea.

> Is the worry having a baseline section and exertion section such a way that one can only parse the baseline section for baseline operation?

Not so much a worry as a way to have a baseline format that can be documented in an RFC and which all flux instances and utilities can understand. Then, for advanced schedulers, etc, the extensions section could be filled with extra information needed for the specific implementation.

> Again, GraphML is an XML; so you don't have to require a special library from the parent as far as the format is formally specified.

Ok, understood. I was possibly confused about how parsing of GraphML works. It would still be nice to keep the number of different encodings down (we'll now have json, yaml, and xml), but for now I'm fine with whatever is most efficient for you.

**Comment by dongahn** [comment]
> Ok, I was worried that boost::graph library or special API from parent instance would be required.

There should be a number of libraries beyond boost that can do this. Probably good idea to document those. I will do that.

> I don't have a good idea. But simple baseline format could be a yaml representation similar to that used in jobspec. This would cut the library requirements down from xml+yaml to just yaml. This is just one idea.

YAML probably works as well. But we will need to extend constructs to support edges and etc. I'm not saying this cannot be done, but like I said, if the graph constructs are already there in GraphML -- tested and hardened with existing libraries, my rational was why reinventing those... For jobspec, YAML makes sense because it needs to be more human readable. But for **R**, this will be likely meant for "machines" to process more than for human users...

> Ok, understood. I was possibly confused about how parsing of GraphML works. It would still be nice to keep the number of different encodings down (we'll now have json, yaml, and xml), but for now I'm fine with whatever is most efficient for you.

I thought about introducing why yet another encoding, xml. But in reality I realized we have already introduced it, as hwloc uses that encoding.

**Comment by grondo** [comment]
> I thought about introducing why yet another encoding, xml. But in reality I realized we have already introduced it, as hwloc uses that encoding.

Yes, but there are no real users of that in flux as yet. (the xml topology is just passed to other hwloc commands)

> YAML probably works as well. But we will need to extend constructs to support edges and etc. I'm not saying this cannot be done, but like I said, if the graph constructs are already there in GraphML -- tested and hardened with existing libraries, my rational was why reinventing those...

Thanks, good points. I don't want to stall your progress with my minor questions. However it has been useful for my understanding, so thanks!

There is also the benefit that there seem to many graph viewers that take in GraphML (e.g. Gephi).

It might help me if we could work through how some simple use cases might work using GraphML from above... e.g. find the correct layout of tasks across *R* for simple task slot shapes (e.g. 1 core, 1 socket, 1 socket, 1 core, etc).

**Comment by dongahn** [comment]
> Thanks, good points. I don't want to stall your progress with my minor questions. However it has been useful for my understanding, so thanks!

It has also been helpful to flesh out more details as well. Thanks!

> It might help me if we could work through how some simple use cases might work using GraphML from above... e.g. find the correct layout of tasks across R for simple task slot shapes (e.g. 1 core, 1 socket, 1 socket, 1 core, etc).

Yeah, I will look more deeply into slot HOWTOs; so I will keep you posted. I had struggled with this under the ECP milestone due coming and punted it by not considering it for that work. So, I need to go back and spend some quality time to work on that logic. I think the logic of finding slots from **R** is its own topic, though. Regardless of what **R** spec we use, finding slots from **R** requires its own effort IMHO.

**Comment by dongahn** [comment]
I believe the results of this discussion is:
- GraphML may be suited for this and R specification in and of itself would be an easier part.
- Need experience on slot support both for the resource matching logic (scheduler) and for tasks layout logic (at the job shell level). 
- Indexing for accelerating finding node-local resources could just be an internal logic of R libraries. 

We agreed that this should be the area of near-term investigation. For now, I call this as "slot support for task layout and resource grouping." This is a co-investigation between execution and scheduling services.

**Comment by dongahn** [comment]
> Need experience on slot support both for the resource matching logic (scheduler) 

FYI -- I have been prototyping slot support using my `resource-query` and at this point I believe this is doable -- essentially first detecting the corresponding resource level for slotting and calling into a special slot-aware subtree walk.

**Comment by SteVwonder** [comment]
Notes from our September meeting about the format for R: https://github.com/flux-framework/rfc/wiki/Brainstorming-on-R

**Comment by dongahn** [comment]
@grondo and @SteVwonder: As we discussed, I will post some R representations to push forward our R spec discussion. At this point, it should be pretty easy for our resource infrastructure to emit a fully concretized `R` with various formats. So we can play with various formats.

### The first obvious idea is to use the resource section format of the jobspec:

Given a tiny machine (called tiny0) and the following jobspec:
![tiny](https://user-images.githubusercontent.com/862123/51155828-bf161080-182d-11e9-9c99-17cd2cf681af.png)

```yaml
version: 1
resources:
    - type: cluster
      count: 1
      with:
        - type: rack
          count: 1
          with:
            - type: node
              count: 1
              with:
                  - type: slot
                    count: 1
                    label: default
                    with:
                      - type: socket
                        count: 1
                        with:
                          - type: core
                            count: 1
                          - type: memory
                            count: 4
```

The resource infrastructure can generate:

```yaml
      - type: cluster
        count: 1
        name: tiny0
        id: 0
        with:
         - type: rack
           count: 1
           name: rack0
           id: 0
           with:
            - type: node
              count: 1
              name: node0
              id: 0
              with:
               - type: socket
                 count: 1
                 name: socket0
                 id: 0
                 exclusive: true
                 with:
                  - type: memory
                    count: 2
                    name: memory1
                    id: 1
                    exclusive: true
                  - type: memory
                    count: 2
                    name: memory0
                    id: 0
                    exclusive: true
                  - type: core
                    count: 1
                    name: core0
                    id: 0
                    exclusive: true
```

A couple Issues:
- `with` implies a multiplicative edge,  and I found that in a fully concretized resource set this semantics can be confusing. What happens if an intermediate vertex's count is multiple? This can lead to an incorrect interpretation of the size of a target vertex. I think introducing the `edge` key whose semantic simply is *associative* would be much cleaner. 

- With the current format, I can't add information that the resource infrastructure uses to manage the graph complexity. For example, the infrastructure allows you to add what subsystem(s)(e.g., containment, network connective and power) a vertex or an edge belongs to. But as is, I can't pass that information and this can affect nested instances. A question is, can we add this optional data into `R` in a way that doesn't increase the overhead of the remote execution system's R parsing much. 

- `slot` and `rank` are missing. As we discussed before, in a fully concretized resource set, it seems better to encode that information as attributes of their resource vertices.  There is an opposite problem in this case. Can we add this info such that it won't affect the overhead of the scheduler's R parsing much. 


Next steps:
Once I get your first feedback, I can show some examples of having `slot` and `rank` added to `R` using resource matching module with hwloc reader.

Let's also discuss what are the implications of adding `edge` key and also optional info key.

**Comment by dongahn** [comment]
I just saw from a next step from @SteVwonder's excellent note: 
> First implement the emitter in resource matching service
>      Emit simple examples as JSON
>     Then work on a reader

I think the JSON representation of the above R looks like:

```json
{
	"type": "cluster",
	"count": 1,
	"name": "tiny0",
	"id": 0,
	"with": {
		"type": "rack",
		"count": 1,
		"name": "rack0",
		"id": 0,
		"with": {
			"type": "node",
			"count": 1,
			"name": "node0",
			"id": 0,
			"with": {
				"type": "socket",
				"count": 1,
				"name": "socket0",
				"id": 0,
				"exclusive": true,
				"with": {
					"type": "socket",
					"count": 1,
					"name": "socket0",
					"id": 0,
					"exclusive": true,
					"with": [{
							"type": "core",
							"count": 1,
							"name": "core0",
							"id": 0,
							"exclusive": true
						},
						{
							"type": "memory",
							"count": 2,
							"name": "memory1",
							"id": 1,
							"exclusive": true
						},
						{
							"type": "memory",
							"count": 2,
							"name": "memory0",
							"id": 0,
							"exclusive": true
						}
					]
				}
			}
		}
	}
}
```

**Comment by grondo** [comment]
@dongahn, thanks for pushing this forward. Here are some thoughts after our quick discussion yesterday. I don't claim to have any answers or great insight, but I want to keep the discussion moving forward.
 
  * Emitting R on JSON is perfect for now. I'm not sure I see a case where yaml would be preferred for R, except in the case of hand edited "resource configuration" input -- however, I assume we would rather offer tools or a higher level DSL to do that anyway.
 * We should plan to use our existing `R_lite` format for now, meanwhile extending the format step-wise with new versions as requirements evolve. To that end, perhaps we want to tweak `R_lite` to include version and or version_name fields?

Some thoughts on your "issues" above:

> with implies a multiplicative edge, and I found that in a fully concretized resource set this semantics can be confusing. What happens if an intermediate vertex's count is multiple? This can lead to an incorrect interpretation of the size of a target vertex. I think introducing the edge key whose semantic simply is associative would be much cleaner.

I trust that you've thought this through better than I, but I don't see how the multiplicative `with` edge has confusing semantics. In a *fully* concretized resource set, you can still emit `with:` links where `count` is `1` and not have the multiplicative property, however allowing `count > 1` for identical resource types in the set would allow for much smaller emitted `R`.

I agree, however, the right apporach might be to start with `edge:` as you suggest, then add special case links like `with:` that are handled in very specific ways. For example, in a fully "concretized" resource set, no resources are *actually* identical, because we are assigning or describing distinct resources. So a multiplicative edge like `with:` would need extra information about how to expand the `count > 1` items into `count` distinct items. E.g. you might have a `idset:` instead of `ids`, and for example, 4 cores on a socket might become:

```json
{ "type": "socket",
   "id": 0,
   "exclusive": false,
   "with": [
      { "type": "core",
         "count": 4,
         "idset": "0-3",
         "exclusive": true
      }
   ]
}
```

A parser of this version of `R` would know to expand this `with:` directive into:

```json
{ "type": "socket",
   "id": 0,
   "exclusive": false,
   "with": [
      { "type": "core",
         "name": "core0",
         "count": 1,
         "id": 0,
         "exclusive": true
      },
      { "type": "core",
         "name": "core1",
         "count": 1,
         "id": 1,
         "exclusive": true
      },
      { "type": "core",
         "name": "core2",
         "count": 1,
         "id": 2,
         "exclusive": true
      },
      { "type": "core",
         "name": "core3",
         "count": 1,
         "id": 3,
         "exclusive": true
      },
   ]
}
```

or whatever the equivalent with `edge:` would be. This is similar to what the original RDL experiment did, and it allowed for very compact representations of thousands of homogeneous nodes. I'm not saying we need this support now, but it would perhaps be a point of future evolution for the R format.

>  A question is, can we add this optional data into R in a way that doesn't increase the overhead of the remote execution system's R parsing much.

What we want is a way to add annotation information to `R`, preferably I think (mostly) outside of the main JSON for R.

We have already talked about allowing generic attributes for resources in `R` as annotation outside of normal R use cases. Initially, what if advanced schedulers used this information to reference resources outside of the main `R` hieararchy. Then the scheduler could embed their extra topology and graph information in one or more completely separate sections of R, which would be considered opaque to anything but that schedulers components. 

This would mean that the top level `R` becomes an object, and the `R` resource format itself would be stored in a well known key within this object (say `resources:`). Any other emitter of `R` could embed a new key with extra information which would be ignored by most parsers, e.g.
```json
{
  "resources": [
    {
      "type": "cluster",
      "name": "test",
      "id": 0,
      "with": [
        {
          "type": "node",
          "name": "node",
          "id": 112,
          "with": [
            {
              "type": "socket",
              "name": "socket",
              "id": 0,
              "with": [
                {
                  "type": "core",
                  "name": "core",
                  "id": 1
                }
              ]
            }
          ]
        }
      ]
    }
  ],
  "sched": {
    "grug": "xml string..."
  }
}
```

The `sched.grug` xml could possibly reference resources from `resources` array either by first embedding uuids into each resource, e.g. `"attributes": { "uuid": ...  }` and reference unique resources that way, or resources could be back referenced some other way, e.g. perhaps by the resource uri something like `cluster0/node112/socket0/core1`. There are probably lots of other solutions as well.

> slot and rank are missing. 

This one requires more thought. `slot` and `rank` would be used directly by flux-core services like the execution system and job shell, so they should be encoded as first-class members of R. We'd have to think through if these items are encoded better in the `resources` section directly or if it might be easier in some separate section of `R`. I don't have any good ideas here.

**Comment by dongahn** [comment]
@grondo: Sorry for the late response. Let me try to reason about this one by one. 

> I trust that you've thought this through better than I, but I don't see how the multiplicative with edge has confusing semantics. In a fully concretized resource set, you can still emit with: links where count is 1

Maybe I'm overthinking this, but I think this can be confusing if we allow a resource pool vertex with `count > 1` to be an intermediate vertex. Say you want to represent 2 compute nodes as a resource pool under which you have 76 cores. A graph can certainly model this: One vertex with two compute nodes aggregated as a pool; Then, from this vertex you have 76 out-edges to `core` vertices. 

Now, if your jobspec is,

```yaml
version: 1
resources:
    - type: cluster
      count: 1
      with:
        - type: rack
          count: 1
          with:
            - type: slot
              count: 1
              label: default
              with:
                - type: node
                  count: 2
                  with:
                    - type: core
                      count: 1
```

And if you emit vertices and edges in a most simplistic way, you would get

```yaml
      - type: cluster
        count: 1
        name: tiny0
        id: 0
        with:
         - type: rack
           count: 1
           name: rack0
           id: 0
           with:
            - type: node
              count: 2
              name: node1
              id: 1
              exclusive: true
              with:
               - type: core
                 count: 1
                 name: core71
                 id: 71
                 exclusive: true
               - type: core
                 count: 1
                 name: core70
                 id: 70
                 exclusive: true
```

But given `with` being multiplicative, I think this can become ambiguous to interpret. In this case, `With` should used interpreted as `associative`, that is the semantics of `edge`. We can certainly mandate `with` in the concretized graph shall be associative. But then in preparation for when we support compression, IMHO we will certainly want `with` to maintain its original `multiplicative` semantics.

**Comment by dongahn** [comment]
> In a fully concretized resource set, you can still emit with: links where count is 1 and not have the multiplicative property

Yes, I agree. In this case, since multiply-by-1 is the same as being associative, this should be okay. It is just that a full concretization is only possible when you had no coarsening in the scheduler's graph data, which in generally cannot be assumed. We can require it in Flux's resource data model but I am unclearly it is a good idea. The high-end systems in a distant future are headed towards the concept of "aggregated resources" where they no longer have the concept of "real" compute nodes...

**Comment by dongahn** [comment]
> I agree, however, the right approach might be to start with edge: as you suggest, then add special case links like with: that are handled in very specific ways.

I completely agree with you! I will propose the `edge` key somewhere so that I can use `edge` for the initial R. I think @trws once proposed this as part of the canonical jobspec, so I can look at the past commits to retrieve and review it.  Seems we agreed that we want to keep the `multiplication` property of `with` and use it as a way to condense `R`.

**Comment by dongahn** [comment]
> A question is, can we add this optional data into R in a way that doesn't increase the overhead of the remote execution system's R parsing much.
What we want is a way to add annotation information to R, preferably I think (mostly) outside of the main JSON for R.

This sounds reasonable. One thing is, though, these extra data are currently not that much of information so not sure if there will be much benefits to have a separate section at least *at this point*. 

Why don't I generate a few examples where those `scheduler` specific data are directly emitted into each vertex and edge as their "attributes" and further our discussions?

**Comment by dongahn** [comment]
> slot and rank are missing.

This one requires more thought. slot and rank would be used directly by flux-core services like the execution system and job shell, so they should be encoded as first-class members of R. We'd have to think through if these items are encoded better in the resources section directly or if it might be easier in some separate section of R. I don't have any good ideas here.

Again some examples seem to help further our discussions. In those future examples, let me emit all of them into first-class members of R.

**Comment by dongahn** [comment]
Seems the next steps should be:

- Propose the edge key
- Have an load option to the resource matching service to generate `R_lite`. This is needed for the initial support for our new remote execution service although some extension will be required;
- As part of that, add other emitter types: JSON and YAML;
- Examples with `slot`, `rank` and scheduler data directly embedded in `R` for further evaluation

**Comment by dongahn** [comment]
@grondo: We didn't have whole lot on the edge key back then. A [commit](https://github.com/flux-framework/rfc/commit/c2c179b4915da92e311a2e5b98c70c6e1927cc06#diff-826e4ce9650989be2fc99c09faae7d16) only has it: 
```
*edge*::	
	**XXX**: need specification for other "edge match descriptors"
```

My initial thought: perhaps we can define the `edge` key as:
```
"dflt_edge_attr": { “subsystem”: string, “relationship”: string }

"edge":
    ?"attr": { “subsystem”: string, “relationship”: string }
    ?"in": ( $vertex_label )
    "out": ( $vertex_label | $resource_vertex )
```

`edge`:  The `edge` key SHALL indicate an edge from a resource vertex to another resource.

If the `in` key is present within it, this SHALL be an edge from the resource vertex referred to by its vertex label to the resource vertex captured by the `out` key. If `in` is omitted, the resource vertex containing the `edge` key is assumed to be the `in` vertex. 

The `out` key SHALL either refer to the destination vertex with a vertex label or be a list conforming to the resource vertex specification. For latter, the each resource vertex appears in this list is assumed to have an edge of the same type from the `in` vertex. 

If `attr` is present, it will describe the subsystem to which this edge belongs and the relationship between two connecting resource vertices. If omitted, it will inherit the default edge attributes. 

With something like this, `R` in a typical case can look like:

```yaml
dflt_edge_attr: { subsystem: containment, relationship: contains }
resource:
  - type: rack
    count: 1
    id: 0
    edge:
      out:
        - type: node
          name: node7
          id: 7
          count: 1
          edge:
            out:
              - type: socket
                count: 1
                edge:
                  out:
                    - type: core
                      name: core0
                      id: 0
                    - type: memory
                      name: memory0
                      count: 4
                      unit: GB

```

**Comment by dongahn** [comment]
> For latter, the each resource vertex appears in this list is assumed to have an edge of the same type from the in vertex.

BTW, if a resource vertex has out edges of different types to other resources, we will have to end up emitting the edge key multiple times from the same `in` vertex. Are duplicate keys allowed in both YAML and JSON?

**Comment by grondo** [comment]
> Are duplicate keys allowed in both YAML and JSON?

I believe keys in *both* YAML and JSON have to be unique. 

> we will have to end up emitting the edge key multiple times from the same in vertex. 

I'm not sure I understand. `in:` and `out:` are valid keys for `edge:` (seems like it could better be called `edges:`), and both are lists, then you can specify any number of `in` or `out` edges for a given resource vertex?

We shouldn't bend over backwards to make the resources section from jobspec fit our `R` though. If it would be better to first emit vertices, then edges in a separate object, perhaps we should allow for that.

**Comment by grondo** [comment]
> "dflt_edge_attr": { “subsystem”: string, “relationship”: string }

I suggest we don't add top-level keys in `R` like this. For readabillity, extensibility (and a bit of sanity), I'd suggest something namespaced, like

```yaml
defaults:
  edge:
     attrs: { "subsystem": "containment", "relationship": "contains" {
```

**Comment by dongahn** [comment]
> I'm not sure I understand. in: and out: are valid keys for edge: (seems like it could better be called edges:), and both are lists, then you can specify any number of in or out edges for a given resource vertex?

Great idea. Let me play with it. Love the idea of `edges` plural.

> We shouldn't bend over backwards to make the resources section from jobspec fit our R though. If it would be better to first emit vertices, then edges in a separate object, perhaps we should allow for that.

Yes. We almost have to allow this. It just that I also wanted to make the similar structure of jobspec's resource section also a valid R format.

**Comment by dongahn** [comment]
> I suggest we don't add top-level keys in R like this. For readabillity, extensibility (and a bit of sanity), I'd suggest something namespaced, like

Yup! I was thinking  along the same line.

**Comment by dongahn** [comment]
> Great idea. Let me play with it. Love the idea of edges plural.

```yaml
defaults:
    edge:
        attrs: { subsystem: containment, out: contains, in: in }
    
resource:
  - type: rack
    count: 1
    id: 0
    edges:
      - out:
        - type: node
          name: node7
          id: 7
          count: 1
          edges:
            - out:
              - type: socket
                count: 1
                edges:
                  - out:
                      - type: core
                        name: core0
                        count: 1
                        id: 0
                      - type: memory
                        name: memory0
                        count: 4
                        unit: GB
                  # how to annotate different out-edge type
                  - out:
                      type: foo
                      name: foo1
                      count: 1
```

I like this direction. But I am not clear what is the best way to annotate an out edge when it has a different attribute then the default. Now I remember I used the singular `edge` key because of this. @grondo: any idea?

**Comment by dongahn** [comment]
```yaml
edges:
    - out:
        ? attrs: {}
        vtx: $resource_vertex
```

We can also do it this way at the expense of being verbose...of course.

**Comment by trws** [comment]
In another thread we had other ways to specify edges, this is a version I had for a relatively dense way to hand-write edges for example:

```yaml
type: node
<power:
  - type: pdu
>with: core
<with: rack
```

Where the prefix characters on the key represented either <in or >out.  That's rather... opaque I would say, but it's OK as shorthand.  If the goal is to do multiple edge types under a single edges key, what we were talking about in 2015 has an example [here](https://github.com/flux-framework/flux-core/issues/354#issuecomment-136518238) with description, copied below, called links back then:

```yaml
    type: Core
    count: 1
    tasks: 1 # defaults to one, meaning one of these rspecs per task, to get one total, use *
    sharing: exclusive
    contains: []
    links:
      - type: uses
        direction: out
        target: 17
```

We discussed this concept and expressing it at length in that issue.

**Comment by trws** [comment]
Happy to, sorry it took so long actually, OpenMP F2F meeting this week.  I like the idea of doing "edges: type : ..." by the way, that's a nice way to handle multiple types without all the extra verbosity of having a "type: ..." key on every one.

**Comment by dongahn** [comment]
>> We shouldn't bend over backwards to make the resources section from jobspec fit our R though. If it would be better to first emit vertices, then edges in a separate object, perhaps we should allow for that.

> Yes. We almost have to allow this. It just that I also wanted to make the similar structure of jobspec's resource section also a valid R format.

Throwing one crazy idea out there: it seems a generic graph format "standard" for JSON is also emerging (like graphml being a small subset of xml to specify a graph), and one possibility is to latch onto that format instead of reinventing our own... http://jsongraphformat.info

**Comment by trws** [comment]
I actually used one of those as the serialization format for the test implementation of the original python version of jobspec. As long as the graph format supports directed multigraphs it would be perfectly reasonable to use it. Asking users to write it is another issue, but using it for passing graphs around would be fine, and could even store “canonicalized” jobspec for that matter.

On 29 Jan 2019, at 13:38, Dong H. Ahn wrote:

Also https://github.com/jsongraph/json-graph-specification

**Comment by dongahn** [comment]
> I actually used one of those as the serialization format for the test implementation of the original python version of jobspec. As long as the graph format supports directed multigraphs it would be perfectly reasonable to use it.

Oh. Good. I am looking at this, and it seems to have all the constructs I need. I will play with it some more then.

> Asking users to write it is another issue

Right. This isn’t so human friendly. The current resource section of jobspec would be far better for that purpose. I don’t know if our requirements include R to be generated directly by human users.

> but using it for passing graphs around would be fine, and could even store “canonicalized” jobspec for that matter.

Yes for the future when the jobspec support a full graph, this could be useful as well.

Dong

**Comment by dongahn** [comment]
OK. It looks like JSON Graph Format (JGF) gives me everything that I need to emit the resource graph. See below an example in JGF to encode the graph in https://github.com/flux-framework/rfc/issues/109#issuecomment-458651472

Pros:
- If JGF can become a standard we should have a plenty of tools we can use -- e.g., validators, reader/writers, visualizers (already things like [jsongraph.py](https://github.com/jsongraph/jsongraph.py) exists) and editors for debugging and human interactivity);
- Similarly, building on a standard may allow us to leverage future techniques such as compression which the community may work on (or we can work on and contribute)

Cons:
- A bit verbose (I will look at the spec to see if I can create a default attributes so that we can optimize. )
- Lose the human-friendly structure as jobspec's resource section has

```json
{  
  "graph":{  
    "nodes":[  
      {  
        "id":"0",
        "metadata":{  
          "type":"rack",
          "name":"rack0",
          "id":0
        }
      },
      {  
        "id":"1",
        "metadata":{  
          "type":"node",
          "name":"node7",
          "id":7
        }
      },
      {  
        "id":"2",
        "metadata":{  
          "type":"socket",
          "name":"socket0",
          "id":0
        }
      },
      {  
        "id":"3",
        "metadata":{  
          "type":"core",
          "name":"core0",
          "id":0
        }
      },
      {  
        "id":"4",
        "metadata":{  
          "type":"memory",
          "name":"memory0",
          "count":2,
          "unit":1073741824,
          "id":0
        }
      },
      {  
        "id":"5",
        "metadata":{  
          "type":"foo",
          "name":"foo0",
          "id":0
        }
      }
    ],
    "edges":[  
      {  
        "source":"0",
        "target":"1",
        "metadata":{  
          "subsystem":"containment",
          "relationship":"contains"
        }
      },
      {  
        "source":"1",
        "target":"2",
        "metadata":{  
          "subsystem":"containment",
          "relationship":"contains"
        }
      },
      {  
        "source":"2",
        "target":"3",
        "metadata":{  
          "subsystem":"containment",
          "relationship":"contains"
        }
      },
      {  
        "source":"2",
        "target":"4",
        "metadata":{  
          "subsystem":"containment",
          "relationship":"contains"
        }
      },
      {  
        "source":"2",
        "target":"5",
        "metadata":{  
          "subsystem":"foo",
          "relationship":"bars"
        }
      }
    ]
  }
}
```

**Comment by dongahn** [comment]
> A bit verbose (I will look at the spec to see if I can create a default attributes so that we can optimize. )

From the current spec, it wasn't clear if JGF has support for adding default node/edge properties. But since this is JSON, we can always add those properties as an extra data (building on @grondo's idea at https://github.com/flux-framework/rfc/issues/109#issuecomment-458664871).

```json
{
    "properties": {
       "edges": [
           {
               "id": "default",
               "subsystem": "containment",
               "relationship": "contains"
           },
           {
               "id": "foo",
               "subsystem": "foo",
               "relationship": "bars"
           }
       ]
    },
   "graph": {
       "nodes": [
            {
                "id": "0",
                 "metadata": {
                     "type": "rack",
                     "name": "rack0",
                     "id": 0 
                  }
            },
            {
                "id": "1",
                "metadata": {
                     "type": "node",
                     "name": "node7",
                     "id": 7
                  } 
            },
            {
                "id": "2",
                "metadata": {
                     "type": "socket",
                     "name": "socket0",
                     "id": 0
                  }
            },
            {
                "id": "3",
                "metadata": {
                     "type": "core",
                     "name": "core0",
                     "id": 0
                  }
            },
            {
                "id": "4",
                "metadata": {
                     "type": "memory",
                     "name": "memory0",
                     "count": 2,
                     "unit": 1073741824,
                     "id": 0
                  }
            },
            {
                "id": "5", 
                "metadata": {
                     "type": "foo",
                     "name": "foo0",
                     "id": 0
                  }
            }
        ],
        "edges": [
            {
                "source": "0",
                "target": "1"
            },
            {
                "source": "1",
                "target": "2"
            },
            {
                "source": "2",
                "target": "3"
            },
            {
                "source": "2",
                "target": "4"
            },
            {
                "source": "2",
                "target": "5",
                "metadata": {
                    "property": "foo"
                 }
            }
        ]
    }
}
```

@grondo and @trws: thoughts?

**Comment by dongahn** [comment]
> A bit verbose 

Also I found that JGF is much more condense and legible than GraphML (https://github.com/flux-framework/rfc/issues/109#issuecomment-335958389), though!

**Comment by grondo** [comment]
> Cons:
> A bit verbose (I will look at the spec to see if I can create a default attributes so that we can optimize. )
> Lose the human-friendly structure as jobspec's resource section has

Can we propose that the base `R` type still contains a hardware topology ("containment" hierarchy) as `resource:` or some other key, with a simple, more human readable hierarchy-only representation, while flux-sched and other advanced schedulers can extend the base R with a `"graph:"` (and other) sections which can be considered opaque to flux-core components?

Really, what is required from flux-core is the basic containment hierarchy, and an ability to map `R` to ranks and construct an `R_local` (exec system), and map task slots to local resources (job shell). We could also offer simple tools that parse and display the `R` for a job (e.g. in queue listing we might just need to pull out a host list, or in a more detailed listing a user-friendly representation of `R`), or users could dump just the `resource:` section of `R` directly.

Not that I am fully opposed to specifying the format of R as JGF, but at this point it seems like overkill for flux-core components. But I'd hate to add extra complexity elsewhere for only a modicum of simplicity in flux-core.

**Comment by dongahn** [comment]
> Can we propose that the base R type still contains a hardware topology ("containment" hierarchy) as resource: or some other key, with a simple, more human readable hierarchy-only representation, while flux-sched and other advanced schedulers can extend the base R with a "graph:" (and other) sections which can be considered opaque to flux-core components?

I think a two section approach has several advantages. For example, this way, we don't have to include rank and slot info into the "graph" section needed for the nested schedulers.

One disadvantage is, though, R will be on the order of twice as big with the two section approach. That may be okay... Generally, the two section approach would be a bit faster in exchange for more needed space.


One alternative appoarch would be to develop a converter layer that converts the graph into the containment as what you require for the execution service.

With either approach, a question remains what should be the exact format for the `resource` section. Like we discuss above, at least O would avoid use of `with` as the first cut.

**Comment by dongahn** [comment]
How about:

https://github.com/flux-framework/rfc/issues/109#issuecomment-458651472 or similar under the `resource` key and JGF under the `sched` or `graph` key?

For our current sprint, `resource` can have `R_lite++` of course?

**Comment by dongahn** [comment]
@grondo and @SteVwonder: I can start to draft an RFC on a two section proposal as a way to push forward this dialogue further, if you like.

**Comment by grondo** [comment]
That would be a great start! Thanks

**Comment by trws** [comment]
I realize I'm coming to this a bit late, but I'd put in my 2c for keeping it one piece.  The hardware topology should always be walkable by just limiting the graph representation to that kind of edge, and if it's in two formats in two parts at least some of the components will have to work with both.  The job shell has to take R emitted from sched for example, so either sched needs to work with both formats or there has to be one that works for both.  Perhaps it would be good to come up with an API or similar for what the core side wants to talk to, that could understand whatever the format is underneath and provide the appropriate information rather than expecting it to walk the resource spec directly?

**Comment by dongahn** [comment]
@trws: 

Thank you for your thoughts.

I think what @grondo wants is to put what's required by the execution service in one section in an easy-to-use format and the full information pertaining to the scheduler in the second section. This isn't too difficult to do by the scheduler.  I hope that we don't find a situation where any component needs to read both sections.  Clearly this isn't optimal in terms of storage and R producer performance. But it has a consumer performance advantage (the execution system only needs to read one section while the nested scheduler instance doesn't needs to read the data like "rank" and "slots") and the lower complexity in the execution system software. The API approach (or converter approach as I suggested above) would be another excellent way to overcome this issue. But what @grondo seems to concern about is the complexity of designing API at this point, as it will have to be a graph code.

**Comment by trws** [comment]
Fair enough.  I'm all for using the format that makes sense, just pointing out that just because it's stored as a graph (or as a tree) doesn't mean we have to access it that way in terms of the API.  The resource spec format is a graph jammed into a tree format, so is JGF, so there are other options if for some reason it turns out to be complicated to get a simple serialization format.
________________________________

**Comment by dongahn** [comment]
Completely agreed! Good thoughts @trws.

**Comment by dongahn** [comment]
Thank you for the good discussions. We will probably want to refer back to this when we evolve `R` later. 

But for now PR https://github.com/flux-framework/rfc/pull/155 resolved the ticket.
